### PR TITLE
Import firmware from folder of loose NCA files

### PR DESCRIPTION
**Pull request by anpilley**
This PR adds a UI menu entry to install firmware. It'll prompt the user where their firmware dump is, verify that it contains at least one NCA file, clean the existing firmware registered folder, and then copy the new files in, and re-run the firmware check.

It does require keys to already be installed (and yuzu to have been restarted to read them), but it should help simplify the process of installing firmware.

### CODE REVIEW

**Review by t895 — APPROVED** [review]

**Comment by anpilley** [comment]
Okay, hold off on merging this PR for a bit, I'm going to see if I can tack on a firmware integrity check step at the end.

File: src/yuzu/main.cpp [hunk 1]
```diff
@@ -4150,6 +4153,146 @@ void GMainWindow::OnVerifyInstalledContents() {
...
+    if (!ContentManager::AreKeysPresent()) {
+        QMessageBox::information(
+            this, tr("Keys not installed"),
+            tr("Install decryption keys and restart yuzu before attempting to install firmware."));
+        return;
+    }
+
+    QString firmware_source_location =
+        QFileDialog::getExistingDirectory(this, tr("Select Dumped Firmware Source Location"),
+                                          QString::fromStdString(""), QFileDialog::ShowDirsOnly);
```
Comment (FearlessTobi) [inline src/yuzu/main.cpp:4172]: `                                          {}, QFileDialog::ShowDirsOnly);`

File: src/yuzu/main.cpp [hunk 2]
```diff
@@ -4150,6 +4153,146 @@ void GMainWindow::OnVerifyInstalledContents() {
...
+    std::filesystem::path firmware_source_path = firmware_source_location.toStdString();
+    if (!Common::FS::IsDir(firmware_source_path)) {
+        progress.close();
+        return;
+    }
+
+    std::vector<std::filesystem::path> out;
+    const Common::FS::DirEntryCallable callback =
+        [&out](const std::filesystem::directory_entry& entry) {
+            if (entry.path().has_extension() && entry.path().extension() == ".nca")
```
Comment (FearlessTobi) [inline src/yuzu/main.cpp:4203]: should be braced

File: src/yuzu/main.cpp [hunk 3]
```diff
@@ -4150,6 +4153,146 @@ void GMainWindow::OnVerifyInstalledContents() {
...
+
+        if (!VfsRawCopy(firmware_src_vfile, firmware_dst_vfile)) {
+            LOG_ERROR(Frontend, "Failed to copy firmware file {} to {} in registered folder!",
+                      firmware_src_path.generic_string(), firmware_src_path.filename().string());
+            success = false;
+        }
+
+        if (QtProgressCallback(100, 20 + (int)(((float)(i) / (float)out.size()) * 70.0))) {
+            success = false;
+            cancelled = true;
```
Comment (FearlessTobi) [inline src/yuzu/main.cpp:4253]: you can just show the message box and then return early here. no need for the extra variable

File: src/yuzu/main.cpp [hunk 4]
```diff
@@ -4150,6 +4153,146 @@ void GMainWindow::OnVerifyInstalledContents() {
...
+        auto firmware_dst_vfile =
+            firmware_vdir->CreateFileRelative(firmware_src_path.filename().string());
+
+        if (!VfsRawCopy(firmware_src_vfile, firmware_dst_vfile)) {
+            LOG_ERROR(Frontend, "Failed to copy firmware file {} to {} in registered folder!",
+                      firmware_src_path.generic_string(), firmware_src_path.filename().string());
+            success = false;
+        }
+
+        if (QtProgressCallback(100, 20 + (int)(((float)(i) / (float)out.size()) * 70.0))) {
```
Comment (FearlessTobi) [inline src/yuzu/main.cpp:4251]: we don't use c style casts

**Comment by FearlessTobi** [comment]
Whoops, I left the tab open and this was not displayed as merged yet for me.

**Pull request merged**